torchbiggraph/model.py

```diff
@@ -590,6 +589,82 @@ def forward(
         return pos_scores, lhs_neg_scores, rhs_neg_scores
 
 
+class L2Comparator(AbstractComparator):
+
+    def prepare(
+        self,
+        embs: FloatTensorType,
+    ) -> FloatTensorType:
+        return embs
+
+    def batched_cdist_l2(self, x1, x2):
+        x1_norm = x1.pow(2).sum(dim=-1, keepdim=True)
+        x2_norm = x2.pow(2).sum(dim=-1, keepdim=True)
+        res = torch.baddbmm(
+            x2_norm.transpose(-2, -1),
+            x1,
+            x2.transpose(-2, -1),
+            alpha=-2
+        ).add_(x1_norm).clamp_min_(1e-30).sqrt_()
```

Comment:
`clamp_min_` is used to avoid having `sqrt(0)` which cannot be differentiated.

Reply:
Do you have any preference/suggestion for the `eps` value we should use? I picked `1e-30` for no particular reason.